src/vs/workbench/contrib/scm/browser/dirtydiffDecorator.ts

```diff
@@ -244,28 +264,96 @@ class DirtyDiffWidget extends PeekViewWidget {
 		const editorHeightInLines = Math.floor(editorHeight / lineHeight);
 		const height = Math.min(getChangeHeight(change) + /* padding */ 8, Math.floor(editorHeightInLines / 3));
 
-		this.renderTitle(labeledChange.labels);
+		this.renderTitle(labeledChange.label);
 
 		const changeType = getChangeType(change);
 		const changeTypeColor = getChangeTypeColor(this.themeService.getColorTheme(), changeType);
 		this.style({ frameColor: changeTypeColor, arrowColor: changeTypeColor });
 
-		this._actionbarWidget!.context = [diffEditorModel.modified.uri, this.model.changes.map(change => change.change), index];
-		this.show(position, height);
+		const providerSpecificChanges: IChange[] = [];
+		for (const change of this.model.changes) {
+			if (change.label === this.model.changes[this._index].label) {
+				providerSpecificChanges.push(change.change);
+			}
+		}
+		this._actionbarWidget!.context = [diffEditorModel.modified.uri, providerSpecificChanges, index];
+		if (usePosition) {
+			this.show(position, height);
+		}
 		this.editor.focus();
 	}
 
-	private renderTitle(labels: string[]): void {
-		const detail = this.model.changes.length > 1
-			? nls.localize('changes', "{0} - {1} of {2} changes", labels.join(', '), this.index + 1, this.model.changes.length)
-			: nls.localize('change', "{0} - {1} of {2} change", labels.join(', '), this.index + 1, this.model.changes.length);
+	private renderTitle(label: string): void {
+		const providerChanges = this.model.mapChanges.get(label)!;
+		const providerIndex = providerChanges.indexOf(this._index);
 
+		let detail: string;
+		let display: 'none' | 'inherit';
+		if (!this.shouldUseDropdown()) {
+			detail = this.model.changes.length > 1
+				? nls.localize('changes', "{0} - {1} of {2} changes", label, providerIndex + 1, providerChanges.length)
+				: nls.localize('change', "{0} - {1} of {2} change", label, providerIndex + 1, providerChanges.length);
+			display = 'none';
+		} else {
+			detail = this.model.changes.length > 1
+				? nls.localize('multiChanges', "{0} of {1} changes", providerIndex + 1, providerChanges.length)
+				: nls.localize('multiChange', "{0} of {1} change", providerIndex + 1, providerChanges.length);
+			display = 'inherit';
+		}
+		if (this.dropdownContainer?.style) {
```

Comment:
Similarly to the comment about `_titleElement`, `dropdownContainer` should always exist. The only reason why it's typed as `| undefined` is because it's not constructed at `ctor` but in the `render*` functions. I suggest switching to `this.dropdownContainer!` and then you can further simplify the code (no need for a `display` local variable).